plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/RestReindexAction.java

```
@@ -0,0 +1,167 @@
+                 * Extract the parameters that we need from the parser. We could
+                 * do away with this hack when search source has an
+                 * ObjectParser.
+                 */
+                Map<String, Object> source = parser.map();
+                Object index = source.remove("index");
+                if (index != null) {
+                    if (index instanceof List) {
+                        @SuppressWarnings("unchecked")
+                        List<String> list = (List<String>) index;
+                        search.indices(list.toArray(new String[list.size()]));
+                    } else if (index instanceof String) {
+                        search.indices((String) index);
+                    } else {
+                        throw new IllegalArgumentException("Expected index to be a list of a string but was [" + index + ']');
+                    }
+                }
+                Object type = source.remove("type");
+                if (type != null) {
+                    if (type instanceof List) {
+                        @SuppressWarnings("unchecked")
+                        List<String> list = (List<String>) type;
+                        search.types(list.toArray(new String[list.size()]));
+                    } else if (type instanceof String) {
+                        search.types((String) type);
+                    } else {
+                        throw new IllegalArgumentException("Expected index to be a type of a string but was [" + type + ']');
+                    }
+                }
+                XContentBuilder builder = XContentFactory.contentBuilder(parser.contentType());
+                builder.map(source);
+                parser = parser.contentType().xContent().createParser(builder.bytes());
+                context.reset(parser);
+                search.source().parseXContent(parser, context);
+            } catch (IOException e) {
+                // TODO throw a better exception
```

Got it. I can just let the exception bubble and the parser expects it.